include/seastar/coroutine/generator.hh

```diff
@@ -22,573 +22,895 @@
+// * Range Yielding:
+//   The buffered variant supports yielding both individual elements and ranges/slices.
+//   This provides flexibility to yield data in whatever form is most convenient for
+//   the producer, while the generator handles efficient batching internally.
 namespace seastar::coroutine::experimental {
 
-template<typename T, template <typename> class Container = std::optional>
-class generator;
+namespace internal {
 
-/// `seastar::coroutine::experimental` is used as the type of the first
-/// parameter of a buffered generator coroutine.
-///
-/// the value of a `buffer_size_t` specifies the size of the buffer holding the
-/// values produced by the generator coroutine. Unlike its unbuffered variant,
-/// the bufferred generator does not wait for its caller to consume every single
-/// produced values. Instead, it puts the produced values into an internal
-/// buffer, before the buffer is full or the generator is suspended. This helps
-/// to alleviate the problem of pingpong between the generator coroutine and
-/// its caller.
-enum class buffer_size_t : size_t;
+namespace unbuffered {
 
-namespace internal {
+template <typename Yielded> class next_awaiter;
 
-using std::coroutine_handle;
-using std::suspend_never;
-using std::suspend_always;
-using std::suspend_never;
-using std::noop_coroutine;
-
-template<typename T>
-using next_value_t = std::optional<T>;
-
-template <template <typename> class Container, typename T>
-concept Fifo = requires(Container<T>&& c, T&& value) {
-    // better off returning a reference though, so we can move away from it
-    { c.front() } -> std::same_as<T&>;
-    c.pop_front();
-    c.push_back(value);
-    bool(c.empty());
-    { c.size() } -> std::convertible_to<size_t>;
-};
+template <typename Yielded>
+class generator_promise_base : public seastar::task {
+    using yielded_deref_type = std::remove_reference_t<Yielded>;
+    using yielded_decvref_type = std::remove_cvref_t<Yielded>;
+    using value_ptr_type = std::add_pointer_t<Yielded>;
 
-template<typename T>
-concept NothrowMoveConstructible = std::is_nothrow_move_constructible_v<T>;
+protected:
+    // a glvalue yield expression is passed to co_yield as its operand. and
+    // the object denoted by this expression is guaranteed to live until the
+    // coroutine resumes. we take advantage of this fact by storing only a
+    // pointer to the denoted object in the promise as long as the result of
+    // dereferencing this pointer is convertible to the Ref type.
+    std::add_pointer_t<Yielded> _value = nullptr;
 
-template<NothrowMoveConstructible T>
-class generator_unbuffered_promise final : public seastar::task {
-    using generator_type = seastar::coroutine::experimental::generator<T, std::optional>;
-    std::optional<seastar::promise<>> _wait_for_next_value;
-    generator_type* _generator = nullptr;
+protected:
+    std::exception_ptr _exception;
+    std::coroutine_handle<> _consumer;
```

Comment:
Is it mandatory that the consumer be a coroutine?

What happens if someone consumes the generator in a do_for_each(), say?

Reply:
Please add that to the changelog. It's a big change.

Reply:
I wonder if we can lift the limitation.

1. add non-explicit `copy_awaiter::operator future<std::optional<T>>` (and similar for the other awaiter)
2. if a non-coroutine consumes a generator, the awaiter decays into a future which is then awaited

Alternatively, generator::operator() can return an intermediate struct, which then decays into a future (for non-coroutines) or is converted into an awaiter (via promise::await_transform()).

This would help interoperability with Seastar RPC streams and with ScyllaDB mutation readers.